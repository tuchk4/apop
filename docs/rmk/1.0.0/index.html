<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="utf-8">
    <title>Home - Documentation</title>

    <script src="scripts/prettify/prettify.js"></script>
    <script src="scripts/prettify/lang-css.js"></script>
    <!--[if lt IE 9]>
      <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
    <link type="text/css" rel="stylesheet" href="styles/prettify.css">
    <link type="text/css" rel="stylesheet" href="styles/jsdoc.css">
</head>
<body>

<input type="checkbox" id="nav-trigger" class="nav-trigger" />
<label for="nav-trigger" class="navicon-button x">
  <div class="navicon"></div>
</label>

<label for="nav-trigger" class="overlay"></label>

<nav>
    <h2><a href="index.html">Home</a></h2><h3>Namespaces</h3><ul><li><a href="actions.html">actions</a><ul class='methods'><li data-type='method'><a href="actions.html#.clear">clear</a></li><li data-type='method'><a href="actions.html#.toCamelCase">toCamelCase</a></li><li data-type='method'><a href="actions.html#.toSnakeCase">toSnakeCase</a></li></ul></li><li><a href="shortcuts.html">shortcuts</a><ul class='methods'><li data-type='method'><a href="shortcuts.html#.clear">clear</a></li><li data-type='method'><a href="shortcuts.html#.toCamelCase">toCamelCase</a></li><li data-type='method'><a href="shortcuts.html#.toSnakeCase">toSnakeCase</a></li></ul></li><li><a href="utils.html">utils</a><ul class='methods'><li data-type='method'><a href="utils.html#.parseArray">parseArray</a></li><li data-type='method'><a href="utils.html#.parseObject">parseObject</a></li><li data-type='method'><a href="utils.html#.reduce">reduce</a></li></ul></li></ul>
</nav>

<div id="main">
    

    



    









    


    <section class="readme">
        <article><h1>RMK</h1><p>Object transformation without mutations.</p>
<pre class="prettyprint source"><code>REMAKE = Function(data): newData
Formula = Function(REMAKE_1 ... REMAKE_n): Function(data): newData</code></pre><p>Pseudo example:</p>
<pre class="prettyprint source lang-js"><code>import rmk from 'rmk';
import toCamelCase from '@rmk/to-camel-case';
import clear from '@rmk/clear';
import update from '@rmk/update';

const formula = rmk(
  toCamelCase(),
  clear(),
  update({
    id: data => data.id + 10
  })
);

const object = [{
  id: 1,
  groupId: 1,
  name: null
}, {
  id: 2,
  groupId: 2,
  name: ''
}];

const newObject = rmk(object);
expect(newObject).toEqual([{
  id: 10,
  group_id: 1
}, {
  id: 20,
  group_id: 2
}])</code></pre></article>
    </section>






</div>

<br class="clear">

<footer>
    Documentation generated by <a href="https://github.com/jsdoc3/jsdoc">JSDoc 3.4.3</a> on Mon Feb 20 2017 17:16:35 GMT+0200 (EET) using the <a href="https://github.com/clenemt/docdash">docdash</a> theme.
</footer>

<script>prettyPrint();</script>
<script src="scripts/linenumber.js"></script>
</body>
</html>