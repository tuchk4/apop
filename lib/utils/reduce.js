import isArray from 'lodash/isArray';
import isPlainObject from 'lodash/isPlainObject';
import parseArray from './parseArray';
import parseObject from './parseObject';

import isMap from 'lodash/isMap';
import isSet from 'lodash/isSet';

/**
 * Parse object or array
 * @memberof utils
 * @param {Object|Array} origin Origin object or array
 * @param {Array} actions Array of transform actions
 *
 * @returns {Object|Array} Transformed object or array by actions
 */
const reduce = (origin, actions) => {
  const arrayLike = isArray(origin) || isSet(origin);
  if (arrayLike) {
    return parseArray(origin, actions);
  }

  const objectLike = isPlainObject(origin) || isMap(origin);
  if (objectLike) {
    return parseObject(origin, actions);
  }

  return origin;
};
export default reduce;
