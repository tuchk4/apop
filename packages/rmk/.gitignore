.DS_Store
/.git
/.idea
/npm-debug.log
/node_modules
/dist
recursive
*.json
/*.js
/*.map
!__tests__
!src
!rollup-build.js
