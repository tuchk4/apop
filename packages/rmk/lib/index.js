import isArray from 'lodash/isArray';
import isObject from 'lodash/isObject';
import flow from 'lodash/flow';

const parseObject = (origin, actions) => {
  return flow(actions)(origin);
};

const parseArray = (origin, actions) => {
  return origin.map(element => parseObject(element, actions));
};

const reduce = (origin, actions) => {
  if (isArray(origin)) {
    return parseArray(origin, actions);
  } else if (isObject(origin)) {
    return parseObject(origin, actions);
  } else {
    throw new Error('Wrong input arguments');
  }
};

const rmk = (...actions) => {
  return origin => reduce(origin, actions);
};
module.exports = rmk;
