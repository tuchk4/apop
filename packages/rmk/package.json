{
  "name": "rmk",
  "version": "2.4.0",
  "description": "Flow transformation data.",
  "main": "index.js",
  "scripts": {
    "prepublish": "yarn build",
    "build": "babel-node ./rollup-build.js",
    "jsdoc2md": "npm run jsdoc2md:actions",
    "jsdoc2md:actions": "jsdoc2md src/actions/*.js  src/actions/**/*.js > docs/ACTIONS.md",
    "jsdoc2md:rmk": "jsdoc2md src/index.js > docs/README.md"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/tuchk4/rmk.git"
  },
  "author": "marvelsrp <marvelsrp@gmail.com>",
  "license": "ISC",
  "bugs": {
    "url": "https://github.com/tuchk4/rmk/issues"
  },
  "jest": {
    "rootDir": "./packages"
  },
  "homepage": "https://github.com/tuchk4/rmk",
  "dependencies": {
    "jsdoc": "^3.4.3",
    "jsdoc-to-markdown": "^3.0.0"
  }
}