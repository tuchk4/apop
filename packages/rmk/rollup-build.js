import { rollup } from 'rollup';

import resolve from 'rollup-plugin-node-resolve';
import uglify from 'rollup-plugin-uglify';
import commonjs from 'rollup-plugin-commonjs';
import babel from 'rollup-plugin-babel';
import { minify } from 'uglify-es';
import babelrc from 'babelrc-rollup';

const babelConfig = {
  presets: [['es2015', { loose: true, modules: false }, 'stage-0']],
  plugins: ['external-helpers', 'transform-object-rest-spread'],
};

var rollUpConfig = {
  format: 'umd',
  sourceMap: 'inline',
  plugins: [
    resolve({
      jsnext: true,
      main: true,
      browser: true,
    }),
    commonjs(),
    babel(
      babelrc({
        addExternalHelpersPlugin: false,
        config: babelConfig,
        exclude: 'node_modules/**',
      })
    ),
  ],
};

async function make(config) {
  config.entry = config.entry || `./src/${config.moduleName}.js`;
  config.dest = config.dest || `./${config.moduleName}.js`;
  config.destMin = config.destMin || `./${config.moduleName}.min.js`;

  let bundle = await rollup({
    entry: config.entry,
    plugins: config.plugins,
  });



  await bundle.write({
    format: config.format,
    moduleName: config.moduleName,
    dest: config.dest,
    sourceMap: config.sourceMap,
  });
  await bundle.write({
    format: config.format,
    moduleName: config.moduleName,
    dest: config.destMin,
    sourceMap: config.sourceMap,
    plugins: [...rollUpConfig.plugins, uglify({}, minify)],
  });
  console.log(`make ${config.dest} and ${config.destMin}`);
}

(async function() {

  await make({...rollUpConfig,
    moduleName: 'rmk',
    entry: './src/rmk.js',
    dest: `./index.js`,
    destMin: `./index.min.js`
  });
  await make({...rollUpConfig,
    moduleName: 'rmkRecursive',
    entry: './src/recursive/rmk.js',
    dest: `./recursive/index.js`,
    destMin: `./recursive/index.min.js`
  });

  const actionBundles = [
    'clear', 'each', 'toCamelCase', 'toSnakeCase', 'rename', 'update',
    'recursive/clear', 'recursive/each',
    'recursive/toCamelCase', 'recursive/toSnakeCase',
    'recursive/rename', 'recursive/update'
  ];

  actionBundles.forEach(async moduleName => await make({...rollUpConfig, moduleName}));

})();
