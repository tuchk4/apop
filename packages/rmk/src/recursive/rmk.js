import rmkActions from './../actions';
import rmkParse from './../utils/parse';

const recursiveActions = {};

Object.keys(rmkActions).forEach(action => {
  recursiveActions[action] = config =>
    rmkParse({
      actions: [rmkActions[action](config)],
      isRecursive: true,
    });
});

const recursive = Object.assign(
  (...actions) => rmkParse({ actions, isRecursive: true }),
  recursiveActions
);

export default recursive;
