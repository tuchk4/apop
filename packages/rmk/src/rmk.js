import rmkActions from './actions';
import rmkParse from './utils/parse';

const singleActions = {};
const recursiveActions = {};

Object.keys(rmkActions).forEach(action => {
  singleActions[action] = config =>
    rmkParse({
      actions: [rmkActions[action](config)],
      isRecursive: false,
    });

  recursiveActions[action] = config =>
    rmkParse({
      actions: [rmkActions[action](config)],
      isRecursive: true,
    });
});

const rmk = Object.assign(
  (...actions) => rmkParse({ actions, isRecursive: false }),
  singleActions
);
const recursive = Object.assign(
  (...actions) => rmkParse({ actions, isRecursive: true }),
  recursiveActions
);

export default Object.assign(rmk, { recursive });
