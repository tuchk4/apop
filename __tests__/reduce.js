import rmk from '../lib';
import reduce from '../lib/utils/reduce';

const before = {
  id: 1,
  first_name: 'Stephen',
  last_name: 'Nelson',
  email: 'snelson0@yellowbook.com',
  gender: 'Male',
  ip_address: '74.191.108.58',
  city: 'Kapotnya',
  phrase: 'Universal 6th generation portal',
  null: null,
  undefined,
  date: new Date(Date.UTC(2001, 1, 1)),
  0: 0,
  1: 1,
  '-1': -1,
  emptyStr: '',
};

test('Object argument', () => {
  const result = reduce(before, [rmk.clear()]);
  expect(result).toMatchSnapshot();
});

test('Array argument', () => {
  const result = reduce([before], [rmk.clear()]);
  expect(result).toMatchSnapshot();
});

test('Map', () => {
  const data = new Map([['x', '0'], ['y', '0'], ['z', null]]);
  const formula = rmk(rmk.clear(), rmk.rename({ x: 'a' }));
  const result = formula(data);
  expect(result).toMatchSnapshot();
});

test('Set with Map', () => {
  const data = new Set();

  data.add(new Map([['x', '0'], ['y', '0']]));
  data.add(new Map([['x', '1'], ['y', '1']]));
  data.add(new Map([['x', null], ['y', '2']]));
  data.add(new Map([['x', '3'], ['y', null]]));
  data.add(new Map([['x', '4'], ['y', '4']]));

  const formula = rmk(rmk.clear());
  const result = formula(data);
  expect(result).toMatchSnapshot();
});
