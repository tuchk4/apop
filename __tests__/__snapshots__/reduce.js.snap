// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Array argument 1`] = `
Array [
  Object {
    "-1": -1,
    "0": 0,
    "1": 1,
    "city": "Kapotnya",
    "date": 2001-02-01T00:00:00.000Z,
    "email": "snelson0@yellowbook.com",
    "first_name": "Stephen",
    "gender": "Male",
    "id": 1,
    "ip_address": "74.191.108.58",
    "last_name": "Nelson",
    "phrase": "Universal 6th generation portal",
  },
]
`;

exports[`Map 1`] = `
Object {
  "a": "0",
  "y": "0",
}
`;

exports[`Object argument 1`] = `
Object {
  "-1": -1,
  "0": 0,
  "1": 1,
  "city": "Kapotnya",
  "date": 2001-02-01T00:00:00.000Z,
  "email": "snelson0@yellowbook.com",
  "first_name": "Stephen",
  "gender": "Male",
  "id": 1,
  "ip_address": "74.191.108.58",
  "last_name": "Nelson",
  "phrase": "Universal 6th generation portal",
}
`;

exports[`Set with Map 1`] = `
Array [
  Object {
    "x": "0",
    "y": "0",
  },
  Object {
    "x": "1",
    "y": "1",
  },
  Object {
    "y": "2",
  },
  Object {
    "x": "3",
  },
  Object {
    "x": "4",
    "y": "4",
  },
]
`;
